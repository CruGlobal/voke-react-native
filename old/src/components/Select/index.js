import React from 'react';
import st from '../../st';
import MultiSelect from 'react-native-multiple-select';
import { isArray } from '../../utils/common';

const Select = ({
  isMulti,
  selectedValue,
  onChange,
  onUpdate,
  options,
  placeholder,
  containerColor,
  isDisabled,
  label,
  ...rest
}) => {
  function handleChange(selectedItems) {
    const selectedOptions = selectedItems.map(
      s => options.find(so => so.value === s) || { value: s },
    );
    const changeVal = isMulti ? selectedOptions : selectedOptions[0];
    onChange && onChange(changeVal);
    onUpdate && onUpdate(changeVal);
  }
  // TODO: Do disabled state

  const arrOfSelected = !selectedValue
    ? []
    : !isMulti && !isArray(selectedValue)
    ? [selectedValue]
    : selectedValue;
  const selectedLabels = arrOfSelected
    .map(s => options.find(so => so.value === s) || { label: s })
    .map(s => s.label);
  return (
    <MultiSelect
      single={!isMulti}
      hideTags={true}
      items={options}
      uniqueKey="value"
      displayKey="label"
      onSelectedItemsChange={handleChange}
      selectedItems={arrOfSelected}
      selectText={placeholder}
      searchInputPlaceholderText={placeholder || 'Search Items...'}
      onChangeInput={text => console.log(text)}
      textInputProps={{ style: [st.pv4, st.w80], autoFocus: false }}
      tagRemoveIconColor={st.colors.black}
      tagBorderColor={st.colors.black}
      tagTextColor={st.colors.black}
      selectedItemTextColor={st.colors.white}
      selectedItemIconColor={st.colors.white}
      itemTextColor={st.colors.white}
      submitButtonColor={st.colors.black}
      submitButtonText="Done Selecting Items"
      // hideDropdown
      styleItemsContainer={[
        { backgroundColor: containerColor || st.colors.white },
      ]}
      // styleInputGroup={[st.pr3]}
      styleSelectorContainer={[st.brbr6, st.brbl6, st.ovh, { marginTop: -5 }]}
      styleListContainer={[]}
      styleRowList={[st.bw1, st.ph4, st.pv5, st.borderTransparent, st.bbBlue]}
      // styleMainWrapper={[st.pr4]}
      // styleDropdownMenu={[st.pr4]}
      styleDropdownMenuSubsection={[st.brbr6, st.brbl6, st.h(50)]}
      styleTextDropdownSelected={[st.pl4, st.tal]}
      styleTextDropdown={[st.pl4, st.pr1]}
      // searchInputStyle={{ flex: 1 }}
    />
  );
};

export default Select;
