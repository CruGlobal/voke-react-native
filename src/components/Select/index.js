import React, { useState } from 'react';

// import MultiSelect from 'react-native-multiple-select';
import { isArray } from 'utils';
import { View, Text, TouchableOpacity } from 'react-native';
import st from 'utils/st';

import Flex from '../Flex';
import VokeIcon from '../VokeIcon';

const Select = ({
  isMulti,
  selectedValue,
  onChange,
  onUpdate,
  options,
  placeholder,
  containerColor,
  isDisabled,
  label,
  ...rest
}) => {
  return (
    <View
      style={{
        width: '100%',
      }}
    >
      {options.map(item => (
        <TouchableOpacity
          onPress={
            isDisabled
              ? () => {}
              : () => {
                  onUpdate(item);
                }
          }
          activeOpacity={0.6}
          accessibilityTraits="button"
          style={{
            paddingVertical: 16,
            paddingHorizontal: 16,
            borderBottomWidth: 0.5,
            borderBottomColor: 'rgba(0,0,0,.15)',
            flex: 1,
            flexDirection: 'row',
          }}
        >
          <Flex
            style={{
              paddingVertical: 6,
              paddingLeft: 10,
              paddingRight: 10,
            }}
          >
            <VokeIcon
              name="checkmark-outline"
              size={18}
              style={{
                color:
                  selectedValue === item.value
                    ? st.colors.orange
                    : st.colors.grey,
                opacity: selectedValue === item.value ? 1 : 0,
              }}
            />
          </Flex>
          <Flex
            value={1}
            style={{
              paddingVertical: 6,
              paddingLeft: 10,
            }}
          >
            <Text
              style={{
                color:
                  selectedValue === item.value
                    ? st.colors.orange
                    : st.colors.darkBlue,
                fontSize: 16,
              }}
            >
              {item.label}
            </Text>
          </Flex>
        </TouchableOpacity>
      ))}
    </View>
  );
};

export default Select;
