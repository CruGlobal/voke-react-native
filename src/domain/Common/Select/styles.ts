import { ReactText } from 'react';
import { Platform } from 'react-native';
import theme from 'utils/theme';
import ui from 'utils/ui';

const styles: { [key: string]: any } = {
	...theme,
	toggle: {
		flexDirection: 'row',
		alignItems: 'center',
	},
	dropdown: {
		position: 'absolute',
		backgroundColor: 'white',
		borderRadius: 4,

		// Shadow
		...Platform.select({
			ios: {
				shadowColor: 'black',
				shadowOffset: { width: 0, height: 2 },
				shadowOpacity: 0.14,
				shadowRadius: 2,
			},
			android: {
				elevation: 8,
			},
		}),
	},

	option: {
		justifyContent: 'flex-start',
		alignSelf: 'center',
		paddingVertical: 10,
		paddingLeft: 10,
		paddingRight: 14,
		flexDirection: 'row',
		width: '100%',
	},
	langOptionText: {
		color: theme.colors.secondary,
		fontSize: theme.fontSizes.m,
	},
	langOptionCheckmark: {
		color: theme.colors.secondary,
		alignSelf: 'center',
		paddingLeft: theme.spacing.s,
		opacity: 0.5,
	},
};

export default styles;
