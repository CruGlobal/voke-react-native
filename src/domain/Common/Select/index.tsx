import Chevron from 'domain/Common/Select/Chevron';

import React, {
  ReactElement,
  RefObject,
  useEffect,
  useRef,
  useState,
} from 'react';
import {
  Modal,
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import Touchable from 'components/Touchable';
import Text from 'components/Text';

import styles from './styles';

interface Option {
  label: string;
  value?: string | number;
  selected?: boolean;
}

interface Props {
  isOpen: boolean; // State managed outside of the component.
  onOpen: () => void;
  onClose: () => void;

  // Toggle/Button options:
  toggleTestId?: string;
  toggleText: string;
  toggleTextStyle?: TextStyle;

  // Selectable options:
  options: Option[]; // Array of options for dropdown.
  optionEl: (option: Option, key: number) => Element; // Custom render template for option elem.
  onSelect: (option: Option, key: number) => void;
}

const Select = (props: Props): ReactElement => {
  const {
    isOpen,
    options,
    onOpen,
    onClose,
    onSelect,
    toggleText,
    toggleTextStyle,
    optionEl,
    toggleTestId,
  } = props;
  // Refs needed for accessing measureInWindow().
  const toggleRef = useRef<TouchableOpacity | null>(null);
  const menuContainerRef: RefObject<View> = useRef(null);
  // Toggle position.
  const [toggleLeft, setToggleLeft] = useState(0);
  const [toggleWidth, setToggleWidth] = useState(0);
  // Dropdown position.
  const [menuTop, setMenuTop] = useState(0);
  const [menuLeft, setMenuLeft] = useState(0);
  const [menuWidth, setMenuWidth] = useState(0);

  useEffect(() => {
    setMenuLeft(toggleLeft + toggleWidth - menuWidth);
  }, [menuWidth, toggleLeft, toggleWidth]);

  const show = (): void => {
    // Get position of the toggle to calculate position for the dropdown.
    toggleRef.current?.measureInWindow((x, y, width, height) => {
      setToggleLeft(x);
      setToggleWidth(width);
      setMenuTop(y + height);
    });
    // Evoke event in the parent component.
    onOpen();
  };

  const hide = (): void => {
    onClose();
  };

  return (
    <View collapsable={false}>
      <TouchableOpacity
        ref={(newRef) => (toggleRef.current = newRef)}
        onPress={(): void => {
          show();
        }}
        style={styles.toggle}
        testID={toggleTestId}
      >
        <Text style={toggleTextStyle}>{toggleText}</Text>
        <Chevron color="white" style={{ marginLeft: 10, marginRight: 2 }} />
      </TouchableOpacity>

      <Modal
        visible={isOpen}
        onRequestClose={hide}
        supportedOrientations={[
          'portrait',
          'portrait-upside-down',
          'landscape',
          'landscape-left',
          'landscape-right',
        ]}
        transparent
      >
        <TouchableWithoutFeedback
          onPress={hide}
          accessible={false}
          testID="dropdownBackdrop"
        >
          <View
            style={[
              StyleSheet.absoluteFill,
              {
                // Don't show menu until menuWidth calculated.
                opacity: menuWidth === 0 ? 0 : 1,
              },
            ]}
          >
            <View
              ref={menuContainerRef}
              onLayout={(): void => {
                menuContainerRef.current?.measureInWindow((x, y, width) => {
                  setMenuWidth(width);
                });
              }}
              style={[
                styles.dropdown,
                {
                  left: menuLeft,
                  top: menuTop,
                },
              ]}
            >
              {options.map((option, key) => {
                return (
                  <Touchable
                    onPress={(): void => {
                      onSelect(option, key);
                      hide();
                    }}
                    style={[
                      styles.option,
                      {
                        backgroundColor: option?.selected
                          ? 'rgba(0,0,0,.07)'
                          : 'transparent',
                      },
                    ]}
                    key={key}
                  >
                    {optionEl(option, key)}
                  </Touchable>
                );
              })}
            </View>
          </View>
        </TouchableWithoutFeedback>
      </Modal>
    </View>
  );
};

export default Select;
