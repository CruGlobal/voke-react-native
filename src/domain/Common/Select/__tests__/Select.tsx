/**
 * @format
 */

import { Text } from 'react-native';
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

// import { toHaveStyle } from '@testing-library/jest-native'; - causes TS error.
// https://github.com/testing-library/jest-dom/issues/123
import '@testing-library/jest-native';
import Select from '../index';

const toggleLabel = 'Toggle Text Test';
const testOptions = [
  { label: 'Option Label 1' },
  { label: 'Option Label 2' },
  {
    label: 'Option Label 3',
    selected: true,
  },
  { label: 'Option Label 4' },
  { label: 'Option Label 5' },
];

// the modal component is automatically mocked by RN and apparently contains
// a bug which make the modal (and it's children) always visible in the test tree
// this is a hack which fix this issue
jest.mock('react-native/Libraries/Modal/Modal', () => {
  const Modal = jest.requireActual('react-native/Libraries/Modal/Modal');
  // @ts-ignore
  return (props) => <Modal {...props} />;
});

beforeEach(() => {
  // https://github.com/facebook/jest/issues/6434#issuecomment-525576660
  jest.useFakeTimers();
});

it('renders correctly', () => {
  const { queryByText, toJSON } = render(
    <Select
      options={testOptions}
      // Toggle button:
      toggleText={toggleLabel}
      toggleTestId="toggleTestId"
      toggleTextStyle={{ color: 'red' }}
      // Selector state change:
      isOpen={false}
      onOpen={() => {}}
      onClose={() => {}}
      onSelect={() => {}}
      // Single option:
      optionEl={(option, key) => {
        return (
          <>
            <Text testID={'singleOption-' + key}>
              {option.label} {option?.selected ? ' - Selected' : ''}
            </Text>
          </>
        );
      }}
    />,
  );

  expect(toJSON()).toMatchSnapshot();
  expect(queryByText(/Toggle Text Test/i)).toBeTruthy();
  expect(queryByText(/Toggle Text Test/i)).toHaveStyle({ color: 'red' });
  expect(queryByText(/Option Label 1/i)).not.toBeTruthy();
});

it('renders correctly when open', async () => {
  const { queryByTestId, queryByText, toJSON } = render(
    <Select
      options={testOptions}
      // Toggle button:
      toggleText={toggleLabel}
      toggleTestId="toggleTestId"
      toggleTextStyle={{ color: 'red' }}
      // Selector state change:
      isOpen={true}
      onOpen={() => {}}
      onClose={() => {}}
      onSelect={() => {}}
      // Single option:
      optionEl={(option, key) => {
        return (
          <>
            <Text testID={'singleOption-' + key}>
              {option.label} {option?.selected ? ' - Selected' : ''}
            </Text>
          </>
        );
      }}
    />,
  );

  expect(queryByText(/Option Label 1/i)).toBeTruthy();
  // Third option should be selected by default (following our test data).
  expect(queryByTestId('singleOption-2')).toHaveTextContent(/Selected/i);
  expect(toJSON()).toMatchSnapshot();
});

it('emits press events', async () => {
  const onPressOpen = jest.fn();
  const onPressClose = jest.fn();
  const onPressSelect = jest.fn();
  const { getByTestId } = render(
    <Select
      options={testOptions}
      // Toggle button:
      toggleText={toggleLabel}
      toggleTestId="toggleTestId"
      // Selector state change:
      isOpen={true}
      onOpen={onPressOpen}
      onClose={onPressClose}
      onSelect={(option, key) => {
        onPressSelect(option, key);
      }}
      // Single option:
      optionEl={(option, key) => {
        return (
          <>
            <Text testID={'singleOption-' + key}>
              {option.label} {option?.selected ? ' - Selected' : ''}
            </Text>
          </>
        );
      }}
    />,
  );

  fireEvent.press(getByTestId('toggleTestId'));
  expect(onPressOpen).toHaveBeenCalled();
  fireEvent.press(getByTestId('dropdownBackdrop'));
  expect(onPressClose).toHaveBeenCalled();
  fireEvent.press(getByTestId('singleOption-0'));
  expect(onPressSelect).toHaveBeenCalledWith({ label: 'Option Label 1' }, 0);
  expect(onPressClose).toHaveBeenCalled();
});
