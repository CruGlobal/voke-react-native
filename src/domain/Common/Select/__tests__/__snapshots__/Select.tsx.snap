// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={false}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
    testID="toggleTestId"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      Toggle Text Test
    </Text>
    <View
      style={
        Array [
          Object {
            "borderLeftColor": "transparent",
            "borderLeftWidth": 0,
            "borderRightColor": "white",
            "borderRightWidth": 1.5,
            "borderTopColor": "white",
            "borderTopWidth": 1.5,
            "height": 7,
            "transform": Array [
              Object {
                "translateX": 1.5,
              },
              Object {
                "translateY": -2.4,
              },
              Object {
                "rotate": "135deg",
              },
            ],
            "width": 7,
          },
          Object {
            "marginLeft": 10,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`renders correctly when open 1`] = `
<View
  collapsable={false}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
    testID="toggleTestId"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      Toggle Text Test
    </Text>
    <View
      style={
        Array [
          Object {
            "borderLeftColor": "transparent",
            "borderLeftWidth": 0,
            "borderRightColor": "white",
            "borderRightWidth": 1.5,
            "borderTopColor": "white",
            "borderTopWidth": 1.5,
            "height": 7,
            "transform": Array [
              Object {
                "translateX": 1.5,
              },
              Object {
                "translateY": -2.4,
              },
              Object {
                "rotate": "135deg",
              },
            ],
            "width": 7,
          },
          Object {
            "marginLeft": 10,
          },
        ]
      }
    />
  </View>
  <RCTModalHostView
    animationType="none"
    hardwareAccelerated={false}
    identifier={1}
    onRequestClose={[Function]}
    onStartShouldSetResponder={[Function]}
    presentationStyle="overFullScreen"
    style={
      Object {
        "position": "absolute",
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
  >
    <View
      collapsable={false}
      style={
        Array [
          Object {
            "flex": 1,
            "left": 0,
            "top": 0,
          },
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          collapsable={true}
          pointerEvents="box-none"
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            accessible={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "opacity": 0,
                },
              ]
            }
            testID="dropdownBackdrop"
          >
            <View
              onLayout={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 4,
                    "position": "absolute",
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.14,
                    "shadowRadius": 2,
                  },
                  Object {
                    "left": 0,
                    "top": 0,
                  },
                ]
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  testID="singleOption-0"
                >
                  Option Label 1
                   
                  
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  testID="singleOption-1"
                >
                  Option Label 2
                   
                  
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "rgba(0,0,0,.07)",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  testID="singleOption-2"
                >
                  Option Label 3
                   
                   - Selected
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  testID="singleOption-3"
                >
                  Option Label 4
                   
                  
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  testID="singleOption-4"
                >
                  Option Label 5
                   
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTModalHostView>
</View>
`;
